jbpm-flow/src/main/java/org/jbpm/process/core/context/variable/CaseVariableInstance.java

```diff
@@ -0,0 +1,102 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.process.core.context.variable;
+
+import java.util.Collection;
+import java.util.Objects;
+
+import org.drools.core.ClassObjectFilter;
+import org.drools.core.common.InternalKnowledgeRuntime;
+import org.jbpm.process.instance.context.variable.VariableScopeInstance;
+import org.kie.api.runtime.KieSession;
+import org.kie.api.runtime.process.CaseData;
+import org.kie.api.runtime.rule.FactHandle;
+
+public class CaseVariableInstance<T> implements VariableInstance<T> {
+
+    private final InternalKnowledgeRuntime knowledgeRuntime;
+    private final Variable variable;
+    private final String caseFileName;
+    private final ValueReference<T> valueReference = new ValueReference<T>() {
+        @Override
+        public T get() {
+            Collection<CaseData> caseFiles = (Collection<CaseData>)
+                    knowledgeRuntime.getObjects(new ClassObjectFilter(CaseData.class));
+            if (caseFiles.size() == 1) {
+                CaseData caseFile = caseFiles.iterator().next();
+                // check if there is case file prefix and if so remove it before checking case file data
+                if (caseFile != null) {
+                    return (T) caseFile.getData(caseFileName);
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public void set(T value) {
+            Collection<CaseData> caseFiles = (Collection<CaseData>)
+                    knowledgeRuntime.getObjects(new ClassObjectFilter(CaseData.class));
+            if (caseFiles.size() == 1) {
+                CaseData caseFile = caseFiles.iterator().next();
+                FactHandle factHandle = knowledgeRuntime.getFactHandle(caseFile);
+                Objects.requireNonNull(factHandle, "factHandle for " + caseFileName + " was null");
+                caseFile.add(caseFileName, value);
+                knowledgeRuntime.update(factHandle, caseFile);
+                ((KieSession) knowledgeRuntime).fireAllRules();
+            }
+        }
+    };
+
+    public CaseVariableInstance(VariableScopeInstance parentScopeInstance, Variable variable) {
+        this.knowledgeRuntime = parentScopeInstance.getProcessInstance().getKnowledgeRuntime();
+        this.variable = variable;
+        if (variable.getName().startsWith(VariableScope.CASE_FILE_PREFIX)) {
+            this.caseFileName = variable.getName().substring(VariableScope.CASE_FILE_PREFIX.length());
+        } else {
+            // should probably issue a warning
```

Comment:
What do you mean by warning?

Reply:
so currently, when a variable begins with `CASE_FILE_PREFIX` we remove the prefix, then look it up as a case variable. Put apparently there are situations (see the `else` branch) where the case variable has **no prefix** and yet we look it up.
https://github.com/kiegroup/jbpm/blob/master/jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java#L52

The procedure in master goes like this:

- look for the given var name in the map of variables
- if no result: see if it's a "magic" name ("processInstanceId" or "parentProcessInstanceId")
- no result: see it's a global
- no result: see if it's a case file -- if there is a prefix, strip it, otherwise just use that name

so in the last case, we just lookup the given name in the case file... even if it's not prefixed. So, the question is: shall we raise a warning? e.g.: "Cannot find $name in scope, looking for $name in case files"

Reply:
what do you guys think about this? Should we log? Should we throw?

Reply:
let's log ... just to stay on safe side for now